win32/src/call_once.c

```diff
@@ -0,0 +1,70 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdint.h>
+#include <stdbool.h>
+
+#include "windows.h"
+
+#include "azure_c_logging/xlogging.h"
+
+#include "azure_c_pal/interlocked.h"
+#include "azure_c_pal/interlocked_hl.h"
+
+#include "azure_c_pal/call_once.h"
+
+MU_DEFINE_ENUM_STRINGS(CALL_ONCE_RESULT, CALL_ONCE_RESULT_VALUES);
+
+/*returns CALL_ONCE_PROCEED if calling code should do proceed to call the code, any other value = code already called*/
+CALL_ONCE_RESULT call_once_begin(volatile_atomic int32_t* state)
+{
+    CALL_ONCE_RESULT result = CALL_ONCE_ALREADY_CALLED; /*optimistically initialize*/
+    
+    int32_t state_local;
+
+    /* 0 - (CALL_ONCE_NOT_CALLED) = not called*/
+    /* 1 - calling... */
+    /* 2 - already called*/
+
+    /*Codes_SRS_CALL_ONCE_02_001: [ call_once_begin shall use interlocked_compare_exchange(state, 1, 0) to determine if user has alredy indicated that the init code was executed with success. ]*/
+    /*Codes_SRS_CALL_ONCE_02_002: [ If interlocked_compare_exchange returns 2 then call_once_begin shall return CALL_ONCE_ALREADY_CALLED. ]*/
+    while ((state_local = interlocked_compare_exchange(state, 1, 0)) != 2)

```

Comment:
>0 [](start = 65, length = 1)

nit: kind of feels like this should be CALL_ONCE_NOT_CALLED here and below

Reply:
I did try to make it better in the comments, but I will change it now.

---
In reply to: [466042417](https://github.com/Azure/azure-c-pal/pull/28#discussion_r466042417) [](ancestors = 466042417)